Diminish container user privs

zancas opened the pull request.

zancas committed:
fix: Run containers as non-root user to avoid permission issues

  - Change default user from root/UID 10003 to container_user/UID 1000
    to match typical host user UID and avoid permission conflicts
  - Update Dockerfile and Dockerfile.ci to create container_user with
    proper home directory and permissions
  - Fix build process to copy binaries through /tmp to avoid permission
    issues when switching between root and container_user
  - Update test task in Makefile.toml to:
    - Run as container_user instead of root
    - Pre-create directories to avoid Docker creating them as root
    - Mount volumes to container_user's home directory
    - Run container in detached mode with proper cleanup handling
  - Add generic docker_cleanup function to base-script for all tasks:
    - Kills child processes on exit
    - Stops containers by ID or name
    - Handles Ctrl-C (SIGINT) and termination signals properly
  - Fix binary permissions in CI images with explicit chmod

  This resolves the issue where 'makers test' created root-owned target
  directories, preventing subsequent builds from writing to them.

handle Ctrl-C with informative message printed once

the script is in the foreground where is handles signals properly, while the container sends output to it from a bg process

partition the container target output into a separate target to a native build

build librocksdb-dev the C++ library for librocksdb-sys the Rust FFI to wrap at build

fix whitespace in Makefile.toml

Created utils/link_to_binaries.sh script that:
  - Takes optional arguments for REPO_ROOT and binary paths
  - Defaults to /home/container_user/zaino for repo root
  - Defaults to /usr/local/bin/ for binary locations
  - Creates symlinks only if binaries don't exist in test_binaries/bins
  - Can be called standalone or used as part of container entrypoint

(cherry picked from commit b726c427c2054e1c35522f2ed9022340308749d8)

  1. Added the utils/link_to_binaries.sh script as the entrypoint for Dockerfile.ci
  2. The script is copied to /usr/local/bin/entrypoint.sh and made executable
  3. Each final target stage now has ENTRYPOINT [/usr/local/bin/entrypoint.sh] and CMD [/bin/bash]
  4. The entrypoint script will check for test binaries and create symlinks if needed, then execute any command passed to the
   container

remove references to PATH from docs

place build artifacts in non-standard location to ensure new uses handle them explicitly

force existence of test_binaries/bins

  The test infrastructure now:
  - Doesn't rely on PATH manipulation
  - Uses explicit paths for binaries in /home/container_user/artifacts
  - Has an entrypoint script that creates symlinks when the container starts
  - Has a comprehensive validation task that tests the entire pipeline

update names and paths

minor cleanups

update help output

hard_code test tag

 rewrote get-ci-image-tag.sh as a Rust script! The new script:
  - Reads the same environment variables (RUST_VERSION, ZCASH_VERSION, ZEBRA_VERSION)
  - Produces identical output to the shell script
  - Is now integrated into all the Makefile.toml tasks that used the shell version

migrate to rust-script for get-ci-image-tag.rs

ensure docker volumes are not root owned prior to docker operations

leave things set as the script dictated

update script name

install rust-script

add workdir for self-hosted runner

revert to correct tag selection for CI image

attempt to run jobs as root on the self-hosted server

attempt to set permissive bits on obstreporous files

I've replaced all actions/checkout calls with direct git commands that:
  1. Initialize a new git repository
  2. Add the remote origin
  3. Fetch only the specific commit SHA
  4. Check out that commit

make a default UID/GID for k8s be 1001, which is overridden locally

revert to zingolabs/dev version of ci.yml

revert attempted CI compatibility hacks

remove offending non-root USER (overwritten by run command locally)

reinstall rust-script

new hack to make CI work hopefully

set up externals stub

run entrypoint, and update TEST_BINARIES_DIR

Fix CI test execution by running entrypoint script to
   setup binaries

  - Run entrypoint script in CI to create symlinks for
  test binaries
  - Set TEST_BINARIES_DIR environment variable for
  tests
  - Add debug logging to diagnose binary path issues
  - Create streamlined test-debug job for faster
  iteration
  - Disable non-essential CI jobs temporarily for
  debugging
  - Remove retries and optimize test execution for
  debug job
  - Fix trailing whitespace in workflow file

nachog00 commented on .github/workflows/ci.yml:
```diff
@@ -169,7 +172,7 @@ jobs:
         uses: test-summary/action@v2
         with:
           paths: "target/nextest/ci/junit.xml"
-        if: always()
```
why disabling this? it's what reports the test results in github's UI

nachog00 commented on Cargo.toml:
```diff
@@ -9,6 +9,7 @@ members = [
     "zaino-fetch",
     "zaino-proto",
 ]
+exclude = [ "externals" ]
```
Maybe this doesn't belong on this PR ? also not an issue

nachog00 commented:
We agreed to have the `DOCKER_DIR` hash not written into `.env.testing-artifacts` but loaded at runtime

nachog00 committed:
Remove Docker directory hash from .env.testing-artifacts
Moving away from "runtime compute of docker related files hash and write into .env.testing-artifacts" towards "just compute it every time"

nachog00 committed:
Refactor Docker hash handling and improve script structure
- Changed `Makefile.toml` to replace `update-docker-hash` with `get-docker-hash`
- Removed hash update logic and added hash retrieval script call in `Makefile.toml`
- Introduced `functions.sh` in `utils/` for shared utility functions, encapsulating the hash calculation
- Updated `get-ci-image-tag.sh` to utilize the newly created `get_docker_hash` function from `functions.sh`
- Created `get-docker-hash.sh` to provide a standalone command for retrieving the Docker hash
- Modified file mode for `helpers.sh` in `utils/` to be executable

These changes:
- Simplify the process by centralizing Docker hash calculation functionality
- Enhance maintainability and reusability of scripts by using shared functions
- Ensure that the Docker hash reflects the current state without manual updates

zancas committed:
remove externals

zancas committed:
reactivate summary

zancas committed:
Merge remote-tracking branch 'zingolabs/diminish_container_user_privs' into diminish_container_user_privs
